src/components/utils/src/socket_win.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2015, Ford Motor Company
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ * Neither the name of the Ford Motor Company nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <windows.h>
+#include <winsock2.h>
+
+#include "utils/socket.h"
+
+namespace utils {
+
+class WinSocketImpl: public Socket::Impl {
+ public:
+  WinSocketImpl(int af, int type, int protocol);
+  explicit WinSocketImpl(SOCKET socket);
+  ~WinSocketImpl();
+
+  bool SetSocketOptions(
+    int level, int optname,
+    const char* optval, size_t optlen) OVERRIDE;
+
+  bool BindSocket(
+    const struct sockaddr* addr, size_t addrlen) OVERRIDE;
+
+  bool ListenSocket(int backlog) OVERRIDE;
+
+  Socket AcceptSocket(
+    struct sockaddr* addr, size_t addrlen) OVERRIDE;
+
+  size_t SendToSocket(
+    const char* buf, size_t length, int flags) OVERRIDE;
+
+ private:
+  SOCKET socket_;
+};
+
+} // namespace utils
+
+
+utils::Socket::Socket(int af, int type, int protocol): impl_(NULL) {
+  impl_ = new WinSocketImpl(af, type, protocol);
+}
+
+utils::WinSocketImpl::WinSocketImpl(int af, int type, int protocol)
+  : socket_(NULL) {
+  socket_ = socket(af, type, protocol);
+  if (INVALID_SOCKET == socket_) {
+    socket_ = NULL;
+  }
+}
+
+utils::WinSocketImpl::WinSocketImpl(SOCKET socket)
+  : socket_(NULL) {
+  if (INVALID_SOCKET == socket) {
+    socket_ = NULL;
+  } else {
+    socket_ = socket;
```

Comment:
I personally prefer explicit if & else. There can the possibility to add some debug log line. Possibly just for debugging.  If & else suites better in such case.